src/network.cpp

```diff
@@ -0,0 +1,1106 @@
+        network_response_callback_t handle_response) {
+    try {
+        ustring response_data = {to_unsigned(response.data()), response.size()};
+        auto parser = ResponseParser(builder);
+        auto result = parser.decrypt(response_data);
+
+        // Process the bencoded response
+        oxenc::bt_list_consumer result_bencode{result};
+
+        if (result_bencode.is_finished() || !result_bencode.is_string())
+            throw std::runtime_error{"Invalid bencoded response"};
+
+        auto response_info_string = result_bencode.consume_string();
+        int16_t status_code;
+        nlohmann::json response_info_json = nlohmann::json::parse(response_info_string);
+
+        if (response_info_json.contains("code") && response_info_json["code"].is_number())
+            status_code = response_info_json["code"].get<int16_t>();
+        else
+            throw std::runtime_error{"Invalid JSON response, missing required status_code field."};
+
+        // If we have a status code that is not in the 2xx range, return the error
+        if (status_code < 200 || status_code > 299) {
+            if (result_bencode.is_finished())
+                return handle_errors(status_code, std::nullopt, info, handle_response);
+
+            return handle_errors(
+                    status_code, result_bencode.consume_string(), info, handle_response);
+        }
+
+        // If there is no body just return the success status
+        if (result_bencode.is_finished())
+            return handle_response(true, false, status_code, std::nullopt, service_node_changes{});
+
+        // Otherwise return the result
+        handle_response(
+                true, false, status_code, result_bencode.consume_string(), service_node_changes{});
+    } catch (const std::exception& e) {
+        handle_response(false, false, -1, e.what(), service_node_changes{});
+    }
+}
+
+template <typename Destination>
+std::vector<onion_path> valid_paths_for_destination(
+        const std::vector<onion_path> paths, const Destination /*destination*/) {
+    return paths;
+}
+
+template <>
+std::vector<onion_path> valid_paths_for_destination(
+        const std::vector<onion_path> paths, const SnodeDestination destination) {
+    std::vector<onion_path> valid_paths = paths;
+    valid_paths.erase(
+            std::remove_if(
+                    valid_paths.begin(),
+                    valid_paths.end(),
+                    [destination](const onion_path& path) {
+                        return std::any_of(
+                                path.nodes.begin(),
+                                path.nodes.end(),
+                                [&destination](const service_node& node) {
+                                    return node == destination.node;
+                                });
+                    }),
+            valid_paths.end());
+    return valid_paths;
+}
+
+template <typename Destination>
+void Network::send_onion_request(
```

Comment:
I tried this but when I did `if constexpr (std::is_same_v<T, SnodeDestination>) {` inside `var::visit(` it was always returning false (whereas if I `if (auto* dest = std::get_if<SnodeDestination>(&destination)) {` it's working correctly) - not sure if this is iOS doing weird stuff (I did see that there are MacOS work-arounds so not sure if they need to also apply to iOS 🤷

Reply:
Sorry for not noticing this earlier: what is `T` here?  Inside a var::visit generic lambda taking a `auto& x` argument, you'd need something along the lines of this to extract the type:
```C++
using T = std::remove_cvref_t<decltype(x)>;
```
to remove any reference and/or const from the type  ("cvref" = const, volatile, reference; volatile is rare but const and reference are quite common), to get the plain, unadorned underlying type.

But since we now are under C++20, using a templated lambda makes this easier:
```C++
[]<typename T>(const T& x) { ... }
```
(so that the `const` and `&` are "outside" the T and thus T will be a plain type).

Reply:
All good - I don't have the original code anymore but it should have been something like this:
```
var::visit([this](const auto& dest){
    using T = std::remove_cvref_t<decltype(dest)>;
    if constexpr (std::is_same_v<T, oxen::quic::RemoteAddress>) 
        ed25519_public_key_.emplace(ed25519_pubkey::from_bytes(dest.view_remote_key()));
    else if constexpr (std::is_same_v<T, ServerDestination>) {
        host_.emplace(dest.host);
        endpoint_.emplace(dest.endpoint);
        protocol_.emplace(dest.protocol);
        method_.emplace(dest.method);

        if (dest.port)
            port_.emplace(*dest.port);

        if (dest.headers)
            headers_.emplace(*dest.headers);
    } else
        throw std::invalid_argument{"Invalid destination type."};
}, destination);
```

I think it was previously always throwing the `Invalid destination type` error, but for some reason it seems to be working now so I'm not sure what was going on...

Since reworking the logic I now only need to deal with this behaviour in the following 3 places:

```
void Builder::set_destination(network_destination destination) {
    ed25519_public_key_.reset();

    if (auto* dest = std::get_if<oxen::quic::RemoteAddress>(&destination))
        ed25519_public_key_.emplace(ed25519_pubkey::from_bytes(dest->view_remote_key()));
    else if (auto* dest = std::get_if<ServerDestination>(&destination)) {
        host_.emplace(dest->host);
        endpoint_.emplace(dest->endpoint);
        protocol_.emplace(dest->protocol);
        method_.emplace(dest->method);

        if (dest->port)
            port_.emplace(*dest->port);

        if (dest->headers)
            headers_.emplace(*dest->headers);
    } else
        throw std::invalid_argument{"Invalid destination type."};
}

std::optional<service_node> node_for_destination(network_destination destination) {
    if (auto* dest = std::get_if<oxen::quic::RemoteAddress>(&destination))
        return *dest;

    return std::nullopt;
}

session::onionreq::x25519_pubkey pubkey_for_destination(network_destination destination) {
    if (auto* dest = std::get_if<oxen::quic::RemoteAddress>(&destination))
        return compute_xpk(dest->view_remote_key());

    if (auto* dest = std::get_if<ServerDestination>(&destination))
        return dest->x25519_pubkey;

    throw std::runtime_error{"Invalid destination."};
}
```